### DISCUSSION TITLE
FreeRTOS-based application running on HART 4 does not 'run' when booting via HSS

### DISCUSSION THREAD

**Discussion by JV73** [polarfire-soc/polarfire-soc-documentation#162] [Bare metal embedded software] [answered]
I created a FreeRTOS-based application running on HART 4 from LIM and verified that it is 'working' via the lim-debug configuration in SoftConsole.  The code base is ~600 kBytes, so it can not boot from eNVM.  I followed the instructions in the series of You-tube videos on Booting a BareMetal application using the HSS to create an HSS package for my application and load it into the eMMC on the ICICLE_KIT. 

The HSS appears to boot the package (see UART_0 boot messages here): 
![image](https://user-images.githubusercontent.com/106761965/181945703-2f0e246f-26fa-4af3-94c4-6fc1d7d86c2e.png)

On UART_1 I see: 
[8.249450] HSS_Boot_PMPSetupHandler(): Hart1 setup complete

I expected to see a reference to HART 4...   I do not see any activity on UART_3 which should be an interactive 'console' for my application.

I expect that I 'missed' something - it looks like the HSS is activating HART 1 rather than HART 4 ?

**Comment by JV73** [comment]
Hm.  Got myself into trouble.  Made a change to the application build to set First Hart to 1 rather than 4, and re-generated the package (bin file).  When I turned on the ICICLE, it showed drive D: and wanted me to format it (instead of showing D, E, F partitions, only one of which D: formatted as the 'boot' partition) - so I did format the D: dive (7.68 Gb) , then I uploaded the new .bin file - and now I get a verification error:
![image](https://user-images.githubusercontent.com/106761965/182007384-c4534733-27ab-4b85-85b0-197b2ed2a0d5.png)

Clearly - I got myself off in the weeds.  Is there a 'recovery procedure' ?

**Reply by JV73** [reply]
OK - I see what I did here.  My brain got into auto-drive mode and I just 'copied' the 'bin' file to 'drive D:' in windows without using the 'usb imager' to load it.  When I updated the MMC using the 'imager' I'm back to the original state - HSS loads the package, but my application does not 'run'.

**Reply by hughbreslin** [reply]
Hey @JV73 you could try creating an "attach" debug configuration for your project to see where things are going wrong when its trying to start - theres a sample attach configuration included with the projects in bare metal examples. This will connect the debugger without resetting the SoC or downloading the app so you'll be able to see where it is and if its trapped :)

**Reply by JV73** [reply]
The 'attach' shows the following: 

![image](https://user-images.githubusercontent.com/106761965/182429330-da3a478b-74eb-448b-a869-6883b958d928.png)

**Reply by JV73** [reply]
I then 'ran' the debugger for a few seconds and I see: 

![image](https://user-images.githubusercontent.com/106761965/182430470-d8452281-b916-47ad-a351-1ca8dd49f585.png)

**Reply by JV73** [reply]
Hm - maybe I have a typo in my application linker script...

**Reply by JV73** [reply]
Tried moving the application to HART 1 and changed the LIM start address to 0x0803_0000 - no difference in behavior - same as above with corresponding change to addresses indicated.

**Reply by JV73** [reply]
main_other_hart() is at 0x0803510 - seems like we 'fell through' somehow and all of the HART are parked at the reset vector.

**Comment by JV73** [comment]
So now that I have my application running on HART 1:

1) if I set the build configuration to LIM-Debug and edit mss_sw_config.h to set FIRST-HART to 0, LAST HART to 1, and IMAGE_LOADED_BY_BOOTLOADER to 0 then build and launch HW_ALL_HARTS_DEBUG.launch ...   I see my application run and interact on UART_3 console interface. 

2) if I then set the build configuration to LIM-Release, and edit mss_sw_config.h to set FIRST-HART to 1, LAST HART to 1, and IMAGE_LOADED_BY_BOOTLOADER to 1 and build; then take the LIM-Release myApp,elf and apply the attached YAML file to package it up, and then use the usbimager to upload the resulting myApp.bin file onto the MMC on the ICICLE kit, and 'reset' the HSS - I see the HSS messages above, and now that I am skipping OpenSBI() I just get 'Hart1 Setup complete' and 'Address to execute 0x8030000 on UART_1.
When I then use SoftConsole and 'attach' the debugger - then 'run' for a bit and then 'pause' - I see  the U54 HARTS waiting for main_hart() and the e51 is off in in the startup code at init_heap()  .clear_l2lim -> config_64_copy -- stepping, it looks like a2 is not decrementing... not sure what's up...  Letting it run for several minutes appears to get nowhere.

**Reply by JV73** [reply]
![image](https://user-images.githubusercontent.com/106761965/182473813-9c7ec735-c1ec-43e2-8c73-870496fd6a86.png)

**Comment by hughbreslin** [comment]
Hi @JV73 just to check:

when the HSS reaches the point where you expect your payload to be running (i.e "Hart1 Setup complete") and you attach the debugger you're seeing the E51 running code from the 0x8xxx_xxxx range? This shouldn't be happening, the E51 should be in the 0xAxxx_xxxx range running the HSS code. 

Does your attach debug configuration -> Startup match the following?

![image](https://user-images.githubusercontent.com/44862881/182583302-8022bc11-4473-4113-be75-a07f2d3a2ee3.png)

I know from your application you'd increased the size of the L2LIM for your app - did you also re-build the HSS with the updated XML for your application? If not the increased LIM size wouldn't be applied at startup as the HSS needs to do it now instead of the HAL in your project.

One thing to point out that has changed since the youtube video you saw: the HSS is no longer running from L2LIM, its running from L2 scratchpad, this means you could run your application from 0x0800_0000 without impacting the HSS, the main thing to keep in mind is that the HSS needs 512K of scratchpad to run.

**Reply by JV73** [reply]
I did re-build the HSS for my application, but it is possible that things got confused somewhere.  I'll run that path. 
The 'hw... attach' debug that I used is the one from the 'mss_rtc' driver example project - which I used as a starting point.   The settings are 'similar' but in the 'run/start/ commands window there is:  thread apply all set $pc=_start

**Reply by nitindeshpande** [reply]
@JV73 
Thats a bug. 
The  'hw... attach' debug launcher should not have "thread apply all set $pc=_start"

**Reply by JV73** [reply]
In the HSS project, there are two .map files in the 'Default' folder.  output-l2scratch.map appears to be the HSS code mapped into the scratchpad.  output-envm-wrapper.map shows things mapped into eNVM, BUT, there 'stack/heap' is in LIM from 800_0000 to 802_4000.

**Reply by JV73** [reply]
Removing the run/start command got me to a different place... 
![image](https://user-images.githubusercontent.com/106761965/182618990-1b4e61b0-d33d-457d-a8f4-4efec8015a3f.png)

**Reply by JV73** [reply]
Tried removing the WFI loop from U54_1.c in my application.  No difference.

**Reply by JV73** [reply]
Hm.  Turns out that the HSS build was using the 'default' MSS configuration - whatever that is.  Re-building the HSS with the XML file that I'm using in the application project resulted in an HSS build that does not run at POR.  I loaded it using the external tool for boot mode 1.  Launching the debug session shows the code stuck in the hss_e51_trap_handler().  So - time to revisit the MSS configurator...

**Reply by JV73** [reply]
Ok - I found that I had 'two' MSS configurations floating around - one with only 256kB for the scratchpad, so I fixed the MSS to use 1024 kB for LIM and 512kB for Scratchpad and re-built both the HSS and the Application with this corrected XML file.   The HSS loads and runs, but the application still is not showing up on UART_3 and the debugger shows:
![image](https://user-images.githubusercontent.com/106761965/182657778-93de7c7c-b57f-4f31-a518-e73c87768674.png)

**Reply by JV73** [reply]
So, I power-cycled the board, pre-empted the HSS by hitting 'enter' and then opened the HSS_workspace and launched the HSS debug session, waited for the 'breakpoint' at the beginning (xA011162), and then started it.   I see the console initialize the DDR and then run the HSS 'normally' until it gets to the prompt '>>'   Messages on UART_1 have changed a little, but it still looks like myApp is not running.  

[152.416955] HSS_U54_Banner(): u54_1: Waiting for E51 instruction
[152.425539] HSS_IHCInit_U54(): Initializing PLIC (Mi-V IHC) for hart 1
[165.07190] HSS_Boot_PMPSetupHandler(): Hart1 setup complete
[165.200268] HSS_GOTO_IPIHandler(): Address to execute is 0x8030000

I paused the debugger - and here's where things are sitting: 

![image](https://user-images.githubusercontent.com/106761965/182660545-6de4ef19-a34a-44b6-9bb4-5b58b87b6be0.png)

**Comment by JV73** [comment]
Ok - now I'm off into a different weed patch.  The HSS is still 'failing to start' HART 1 - even tho it reports that it thinks that it has (see above); but, as I am experimenting with different build flows, I noticed that:

When I set IMAGE_LOADED_BY_BOOTLOADER = 1 to target an HSS package, if I try to use the 'LIM-DEBUG' build configuration, I get a linker error: mss_entry.S:350: undefined reference to `__app_stack_bottom'  and also '__app_stack_top'.  So to build for use with the HSS I MUST use the LIM-Release build configuration which resolves that error.

However: I just noticed in the LIM-Release/myApp.lst file - most of my code is missing.  The .text 'size' has gone from ~370k down to 17k.  I compared the optimization and linker settings between the two and found that 'debug' has optimization set to 'none' and Release used standard gnu11 optimization, so I turned off 'optimization' in the LIM-Release settings - but it did not make any difference - still missing most of the application code.

I can try using a 'debug build' with the HSS, but it will have IMAGE_LOADED_BY_BOOTLOADER = 0 in order to get past the linker error wrt `__app_stack_bottom/top' - which likely causes problems of it's own.  

So what could be 'different' about the LIM-Release build that is causing my application code to evaporate ?

**Reply by JV73** [reply]
Hm - I updated the 'includes' in the LIM-release configuration to match those in the LIM-debug configuration, and now the builds both behave the same - they both get the linker error about __app_stack_bottom/top... so I must have had an extra 'path' included in the LIM-release configuration... which I deleted without noting what it was, but I suspect is was incorrect anyway.

I notice that there is a 'ddr_loaded_by_bootloader.ld' ... do I need an equivalent 'LIM_loaded_by_bootloader.ld' script to provide '__app_stack_bottom/top' pointing to the e51 stack area ?    What am I missing ?

**Reply by hughbreslin** [reply]
Hi Jim,

yes you will need a "LIM-loaded-by-bootloader" config, [Discussion 109](https://github.com/orgs/polarfire-soc/discussions/109) shows how to set it up :)

**Comment by JV73** [comment, marked as answer]
OK - that did it. To Recap:
1) I moved the 'main' application code from HART 4 to HART 1 - likely not necessary, but makes step 2 more straightforward.
2) Modified mss_sw_config.h to set FIRST-HART to 1, LAST HART to 1, and IMAGE_LOADED_BY_BOOTLOADER to 1.
3) Created my own mpfs-lim-loaded-by-boot-loader.ld per Discussion 109 and changed:
     l2lim (rwx) : ORIGIN = 0x08030000, LENGTH = 1024k   /* to avoid conflict with HSS stack area */
     scratchpad(rwx) : ORIGIN = 0x0A000000, LENGTH = 512k  based on changed made in the MSS_Configurator to increase LIM space
     HEAP_SIZE = 64k; 
     STACK_SIZE_U54_APPLICATION = 16k;    /* based on my application's needs */
     added the following at the end of .stack to map this space for FreeRTOS: 
         PROVIDE(__freertos_irq_stack_top = .);
4)  YAML settings for the packager as shown elsewhere herein.
